modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteCorrelatedNestedLoopJoin.java

```diff
@@ -93,14 +96,27 @@ public IgniteCorrelatedNestedLoopJoin(RelInput input) {
 
     /** {@inheritDoc} */
     @Override public List<Pair<RelTraitSet, List<RelTraitSet>>> deriveCollation(RelTraitSet nodeTraits, List<RelTraitSet> inputTraits) {
-        // We preserve left edge collation only if batch size == 1
-        if (variablesSet.size() == 1)
-            return super.deriveCollation(nodeTraits, inputTraits);
-
         RelTraitSet left = inputTraits.get(0), right = inputTraits.get(1);
+        RelCollation leftCollation = TraitUtils.collation(left), rightCollation = TraitUtils.collation(right);
 
-        return ImmutableList.of(Pair.of(nodeTraits.replace(RelCollations.EMPTY),
-            ImmutableList.of(left.replace(RelCollations.EMPTY), right.replace(RelCollations.EMPTY))));
+        if (!isPrefix(rightCollation.getKeys(), joinInfo.rightKeys))
```

Comment:
it might be worth keeping the partial sort as well. In this case, we will be dealing with a large range, but we will save on explicit sorting. See IgniteMergeJoin#maxPrefix and its usage